src/platform/datapath_winrdma.c

```diff
@@ -0,0 +1,5151 @@
+        Socket->Freed = TRUE;
+        CXPLAT_FREE(CxPlatSocketToRaw(Socket), QUIC_POOL_SOCKET);
+    }
+}
+
+//
+// Create an RDMA connection and associate a socket
+//
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+SocketCreateRdmaInternal(
+    _In_ CXPLAT_DATAPATH* Datapath,
+    _In_ CXPLAT_SOCKET_TYPE Type,
+    _In_opt_ const QUIC_ADDR* LocalAddress,
+    _In_opt_ const QUIC_ADDR* RemoteAddress,
+    _In_opt_ void* CallbackContext,
+    _In_ const CXPLAT_RDMA_CONFIG* Config,
+    _Out_ CXPLAT_SOCKET** NewSocket
+    )
+{
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    DWORD BytesReturned = 0;
+    uint16_t PartitionIndex;
+    RDMA_CONNECTION* RdmaConnection = NULL;
+    RDMA_NDSPI_ADAPTER* NdAdapter = NULL;
+    uint8_t *Buffer = NULL;
+    uint8_t *SendRingBuffer = NULL;
+    uint8_t *RecvRingBuffer = NULL;
+    uint8_t *OffsetBuffer = NULL;
+    uint8_t *RemoteOffsetBuffer = NULL;
+    size_t BufferSize = 0;
+
+    *NewSocket = NULL;
+
+    if (!Datapath ||
+        !Datapath->RdmaAdapter ||
+        !Config ||
+        !Config->SendRingBufferSize ||
+        !Config->RecvRingBufferSize ||
+        Config->SendRingBufferSize < MIN_RING_BUFFER_SIZE ||
+        Config->RecvRingBufferSize < MIN_RING_BUFFER_SIZE)
+    {
+        QuicTraceLogError(
+            CreateRdmaSocketFailed,
+            "CreateRdmaSocket failed, invalid address family");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    CXPLAT_DBG_ASSERT(Datapath->RdmaHandlers.Receive != NULL);
+
+    CXPLAT_SOCKET_PROC* SocketProc = NULL;
+    uint32_t RawSocketLength = CxPlatGetRawSocketSize() + sizeof(CXPLAT_SOCKET_PROC);
+    CXPLAT_SOCKET_RAW* RawSocket = CXPLAT_ALLOC_PAGED(RawSocketLength, QUIC_POOL_SOCKET);
+    if (RawSocket == NULL) {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "CXPLAT_SOCKET",
+            RawSocketLength);
+        return QUIC_STATUS_OUT_OF_MEMORY;
+    }
+    CXPLAT_SOCKET* Socket = CxPlatRawToSocket(RawSocket);
+
+    QuicTraceEvent(
+        DatapathCreated,
+        "[data][%p] Created, local=%!ADDR!, remote=%!ADDR!",
+        Socket,
+        CASTED_CLOG_BYTEARRAY(LocalAddress ? sizeof(*LocalAddress) : 0, LocalAddress),
+        CASTED_CLOG_BYTEARRAY(RemoteAddress ? sizeof(*RemoteAddress) : 0, RemoteAddress));
+
+    ZeroMemory(RawSocket, RawSocketLength);
+    Socket->Datapath = Datapath;
+    Socket->ClientContext = CallbackContext;
+    Socket->HasFixedRemoteAddress = TRUE;
+    Socket->RdmaContext = NULL;
+    Socket->Type = Type;
+    Socket->UseRdma = TRUE;
+
+    if (LocalAddress) {
+        CxPlatConvertToMappedV6(LocalAddress, &Socket->LocalAddress);
+    } else {
+        Socket->LocalAddress.si_family = QUIC_ADDRESS_FAMILY_INET6;
+    }
+    PartitionIndex =
+        RemoteAddress ?
+            ((uint16_t)(CxPlatProcCurrentNumber() % Datapath->PartitionCount)) : 0;
+    Socket->Mtu = CXPLAT_MAX_MTU;
+    Socket->RecvBufLen =
```

Comment:
This should probably be updated for RDMA. I don't think receive coalescing can apply